crypto/mlkem/mlkem-variant.c

```diff
@@ -0,0 +1,1215 @@
+    for (i = 0; i < OSSL_MLKEM_SHARED_SECRET_BYTES; i++)
+        out_shared_secret[i] = constant_time_select_8(mask,
+                                                      key_and_randomness[i],
+                                                      failure_key[i]);
+    return 1;
+}
+
+int ossl_mlkem_decap(uint8_t *out_shared_secret,
+                     const uint8_t *ciphertext,
+                     size_t ciphertext_len,
+                     const private_key *priv,
+                     ossl_mlkem_ctx *mlkem_ctx)
+{
+    if (mlkem_ctx == NULL)
+        return 0;
+
+    if (ciphertext_len != CIPHERTEXT_BYTES) {
+        RAND_bytes_ex(mlkem_ctx->libctx, out_shared_secret,
+                      MLKEM_SHARED_SECRET_BYTES, SECBITS);
+        return 0;
+    }
+    return mlkem_decap(out_shared_secret, ciphertext, priv, mlkem_ctx);
+}
+
+/*
+ * mlkem_parse_public_key_no_hash parses |in| into |pub| but doesn't calculate
+ * the value of |pub->public_key_hash|.
+ */
+static int mlkem_parse_public_key_no_hash(public_key *pub, uint8_t *in,
+                                          ossl_mlkem_ctx *mlkem_ctx)
+{
+    if (vector_decode(&pub->t, in, kLog2Prime) != 1)
+        return 0;
+    memcpy(pub->rho, in + kEncodedVecSize, sizeof(pub->rho));
+    matrix_expand(&pub->m, pub->rho, mlkem_ctx);
+    return 1;
+}
+
+int ossl_mlkem_parse_public_key(public_key *out_public_key, uint8_t *in,
+                                ossl_mlkem_ctx *mlkem_ctx)
+{
+    if (!mlkem_parse_public_key_no_hash(out_public_key, in, mlkem_ctx))
+        return 0;
+    return hash_h(out_public_key->public_key_hash, in,
+                  kEncodedPubkeySize, mlkem_ctx);
+}
+
+static int mlkem_marshal_private_key(uint8_t *out, const private_key *priv)
+{
+    uint8_t *out_curr = out;
+
+    vector_encode(out_curr, &priv->s, kLog2Prime);
+    out_curr += kEncodedVecSize;
+
+    if (ossl_mlkem_marshal_public_key(out_curr, &priv->pub) != 1)
+        return 0;
+    out_curr += kEncodedPubkeySize;
+
+    memcpy(out_curr, priv->pub.public_key_hash, MLKEM_PKHASH_BYTES);
+    out_curr += MLKEM_PKHASH_BYTES;
+
+    memcpy(out_curr, priv->fo_failure_secret, MLKEM_RANDOM_BYTES);
+    return 1;
+}
+
+int ossl_mlkem_marshal_private_key(uint8_t *out, const private_key *priv)
+{
+    return mlkem_marshal_private_key(out, priv);
+}
+
+int ossl_mlkem_parse_private_key(private_key *priv, uint8_t *in,
+                                   ossl_mlkem_ctx *mlkem_ctx)
+{
+    uint8_t *in_curr = in;
+
+    if (!vector_decode(&priv->s, in_curr, kLog2Prime))
+        return 0;
+    in_curr += kEncodedVecSize;
+
+    if (!mlkem_parse_public_key_no_hash(&priv->pub, in_curr, mlkem_ctx))
+        return 0;
+    in_curr += kEncodedPubkeySize;
+
+    memcpy(priv->pub.public_key_hash, in_curr, MLKEM_PKHASH_BYTES);
+    in_curr += MLKEM_PKHASH_BYTES;
+
+    memcpy(priv->fo_failure_secret, in_curr, MLKEM_RANDOM_BYTES);
+    return 1;
+}
+
+#endif /* OPENSSL_NO_MLKEM */
```

Comment:
Likely needs an `#else` and `NON_EMPTY_TRANSLATION_UNIT` but as noted, this is better done in the build.info